method/conST/conST.py

```diff
@@ -0,0 +1,295 @@
+observation_file = args.observations
+
+if args.neighbors is not None:
+    neighbors_file = args.neighbors
+if args.matrix is not None:
+    matrix_file = args.matrix
+if args.dim_red is not None:
+    dimred_file = args.dim_red
+if args.image is not None:
+    image_file = args.image
+if args.config is not None:
+    with open(args.config, "r") as f:
+        config = json.load(f)
+
+n_clusters = args.n_clusters
+technology = args.technology
+seed = args.seed
+
+# ... or AnnData if you want
+def get_anndata(args):
+    import anndata as ad
+    import numpy as np
+    import pandas as pd
+    import scipy as sp
+    from PIL import Image
+
+    observations = pd.read_table(args.observations, index_col=0)
+    features = pd.read_table(args.features, index_col=0)
+
+    coordinates = (
+        pd.read_table(args.coordinates, index_col=0)
+        .loc[observations.index, :]
+        .to_numpy()
+    )
+
+    adata = ad.AnnData(obs=observations, var=features, obsm={"spatial": coordinates})
+
+    if args.matrix is not None:
+        X = sp.io.mmread(args.matrix)
+        if sp.sparse.issparse(X):
+            X = X.tocsr()
+        adata.X = X
+
+    if args.neighbors is not None:
+        adata.obsp["spatial_connectivities"] = sp.io.mmread(args.neighbors).T.tocsr()
+
+    # Filter by selected samples/features
+    if "selected" in adata.obs.columns:
+        adata = adata[observations["selected"].astype(bool), :]
+    if "selected" in adata.var.columns:
+        adata = adata[:, features["selected"].astype(bool)]
+
+    if args.dim_red is not None:
+        adata.obsm["reduced_dimensions"] = (
+            pd.read_table(args.dim_red, index_col=0).loc[adata.obs_names].to_numpy()
+        )
+
+    if args.image is not None:
+        adata.uns["image"] = np.array(Image.open(args.img))
+    else:
+        adata.uns["image"] = None
+
+    return adata
+
+
+adata = get_anndata(args)
+adata.var_names_make_unique()
+
+
+##############
+### conST ####
+########################################################################################################################
+
+if not os.path.exists(f'{out_dir}'):
+    os.makedirs(f'{out_dir}')
+
+params = args
+
+# Add config to params
+for key, value in config.items():
+    setattr(params, key, value)
+
+# Tool imports
+import random
+import torch
+import numpy as np
+import pandas as pd
+import tempfile
+import anndata
+import matplotlib.pyplot as plt
+import scanpy as sc
+import warnings
+from sklearn import metrics
+warnings.filterwarnings('ignore')
+
+random.seed(seed)
+np.random.seed(seed)
+torch.manual_seed(seed)
+torch.cuda.manual_seed(seed)
+device = 'cuda:0' if torch.cuda.is_available() else 'cpu'
+print('Using device: ' + device)
+params.device = device
+
+os.environ['PYTHONHASHSEED'] = str(seed)
+torch.backends.cudnn.benchmark = False
+torch.backends.cudnn.deterministic = True
+
+# TODO this is what conST uses for data preprocessing
+def adata_preprocess(i_adata, min_cells=3, pca_n_comps=300):
```

Comment:
If they run this with 300 PCs we should probably keep this separate from the preprocessing, I don' think we will scale the numbers of PCs that high. Also I am sceptical about how much information these PCs are adding.